File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    // Write your first then lower case function here
    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream()
                .filter(predicate)  // Filter elements based on the given predicate
                .findFirst()        // Find the first matching element
                .map(String::toLowerCase);  // Convert the result to lowercase if present
    }
    // Write your say function here
    public static Say say() {
        return new Say("");
    }

    // say method with a single argument
    public static Say say(String word) {
        return new Say(word);
    }

    public static class Say {
        private final String phrase;

        public Say(String word) {
            this.phrase = word;
        }

        public Say and(String word) {
            return new Say(this.phrase + " " + word);
        }

        public String phrase() {
            return phrase;
        }
    }

    // Write your line count function here
    public static long meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            return reader.lines()
                    .filter(line -> !line.trim().isEmpty() && !line.trim().startsWith("#"))
                    .count();
        }
    }
}

// Write your Quaternion record class here
record Quaternion(double a, double b, double c, double d) {
    
    public static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public static final Quaternion I = new Quaternion(0, 1, 0, 0);
    public static final Quaternion J = new Quaternion(0, 0, 1, 0);
    public static final Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if(Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public Quaternion times(Quaternion other) {
        return new Quaternion(
            (this.a*other.a() - this.b*other.b() - this.c*other.c() - this.d*other.d()),
            (this.a*other.b() + this.b*other.a() + this.c*other.d() - this.d*other.c()),
            (this.a*other.c() - this.b*other.d() + this.c*other.a() + this.d*other.b()),
            (this.a*other.d() + this.b*other.c() - this.c*other.b() + this.d*other.a())
        );
    }

    public Quaternion plus(Quaternion other) {
        return new Quaternion(
            a + other.a(),
            b + other.b(),
            c + other.c(),
            d + other.d()
        );
    }


    public List<Double> coefficients() {
        return List.copyOf(List.of(this.a, this.b, this.c, this.d));
    }

    public Quaternion conjugate() {
        return new Quaternion(
            a, -1*b, -1*c, -1*d
        );
    }

    public String toString() {
        String ret = "";
        if (a==0 && b==0 && c==0 && d==0) {
            return "0";
        } 

        if (a != 0) {
            ret += Double.toString(a);
        }

        if (b != 0) {
            if(ret.equals("")) {
                if(b == 1) {
                    ret += "i";
                } else if (b == -1) {
                    ret += "-i";
                } else {
                    ret += Double.toString(b) + "i";
                }
            } else {
                if (b > 0) {
                    ret += "+";
                }

                if(Math.abs(b) != 1) {
                    ret += Double.toString(b) + "i";
                } else {
                    ret += "i";
                }
            }
        }

        if (c != 0) {
            if(ret.equals("")) {
                if(c == 1) {
                    ret += "j";
                } else if (c == -1) {
                    ret += "-j";
                } else {
                    ret += Double.toString(c) + "j";
                }
            } else {
                if (c > 0) {
                    ret += "+";
                }

                if(Math.abs(c) != 1) {
                    ret += Double.toString(c) + "j";
                } else {
                    ret += "j";
                }
            }
        }

        if (d != 0) {
            if(ret.equals("")) {
                if(d == 1) {
                    ret += "k";
                } else if (d == -1) {
                    ret += "-k";
                } else {
                    ret += Double.toString(d) + "k";
                }
            } else {
                if (d > 0) {
                    ret += "+";
                }

                if(Math.abs(d) != 1) {
                    ret += Double.toString(d) + "k";
                } else {
                    ret += "k";
                }
            }
        }
        return ret;
    }
}
// Write your BinarySearchTree sealed interface and its implementations here
abstract sealed class BinarySearchTree permits Empty, Node {

    abstract int size();
    abstract boolean contains(String value);
    abstract BinarySearchTree insert(String value);

    @Override
    public abstract String toString();
}

final class Empty extends BinarySearchTree {

    public static final Empty INSTANCE = new Empty();

    // Private constructor to ensure a singleton pattern
    public Empty() {}

    @Override
    int size() {
        return 0;
    }

    @Override
    boolean contains(String value) {
        return false;
    }

    @Override
    BinarySearchTree insert(String value) {
        return new Node(value, INSTANCE, INSTANCE); // Return a new Node with Empty children
    }

    @Override
    public String toString() {
        return "()"; // Empty node represented as ()
    }
}

final class Node extends BinarySearchTree {

    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    public Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    int size() {
        // Compute the size by adding 1 for the current node plus the sizes of the left and right children
        return 1 + left.size() + right.size();
    }

    @Override
    boolean contains(String value) {
        if (this.value.equals(value)) {
            return true;
        }
        if (value.compareTo(this.value) < 0) {
            return left.contains(value); // Search the left subtree
        }
        return right.contains(value); // Search the right subtree
    }

    @Override
    BinarySearchTree insert(String value) {
        if (this.value.equals(value)) {
            return this; // If the value is already present, return this node (no duplicates)
        }
        if (value.compareTo(this.value) < 0) {
            // If value is smaller, insert it in the left subtree
            return new Node(this.value, left.insert(value), right);
        }
        // Otherwise, insert it in the right subtree
        return new Node(this.value, left, right.insert(value));
    }

    @Override
    public String toString() {
        // In-order traversal and clean up redundant empty nodes
        return String.format("(%s%s%s)", left, value, right).replace("()", "");
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// first then lower case function
fun firstThenLowerCase(list: List<String>, predicate: (String) -> Boolean): String? {
    return list.firstOrNull(predicate)?.lowercase()
}
// say function 
class Say(private val currentPhrase: String = "") {

    fun and(word: String): Say {
        return Say(this.currentPhrase + " " + word)
    }

    val phrase: String
        get() = currentPhrase
}

fun say(word: String = ""): Say {
    return Say(word)
}

//  meaningfulLineCount function 
fun meaningfulLineCount(filename: String): Long {
    BufferedReader(FileReader(filename)).use { reader ->
        return reader.lineSequence()
            .filter { it.trim().isNotEmpty() }
            .filter { !it.trim().startsWith("#") }
            .count()
            .toLong()
    }
}
//  Quaternion 
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion {
        val newA = a * other.a - b * other.b - c * other.c - d * other.d
        val newB = a * other.b + b * other.a + c * other.d - d * other.c
        val newC = a * other.c - b * other.d + c * other.a + d * other.b
        val newD = a * other.d + b * other.c - c * other.b + d * other.a
        return Quaternion(newA, newB, newC, newD)
    }

    fun coefficients(): List<Double> {
        return listOf(a, b, c, d)
    }

    fun conjugate(): Quaternion {
        return Quaternion(a, -b, -c, -d)
    }

    override fun toString(): String {
        return when {
            a == 0.0 && b == 0.0 && c == 0.0 && d == 0.0 -> "0"
            else -> formatQuaternion(a, b, c, d)
        }
    }

    private fun formatQuaternion(a: Double, b: Double, c: Double, d: Double): String {
        val terms = mutableListOf<String>()

        if (a != 0.0) {
            terms.add(formatTerm(a, "", isFirst = true))
        }
        if (b != 0.0) {
            terms.add(formatTerm(b, "i", isFirst = terms.isEmpty()))
        }
        if (c != 0.0) {
            terms.add(formatTerm(c, "j", isFirst = terms.isEmpty()))
        }
        if (d != 0.0) {
            terms.add(formatTerm(d, "k", isFirst = terms.isEmpty()))
        }

        return terms.joinToString(separator = "")
    }

    private fun formatTerm(value: Double, suffix: String, isFirst: Boolean): String {
        return when {
            value == 0.0 -> ""
            value == 1.0 && suffix.isNotEmpty() -> if (isFirst) suffix else "+$suffix"
            value == -1.0 && suffix.isNotEmpty() -> "-$suffix"
            value > 0 && suffix.isNotEmpty() -> if (isFirst) "$value$suffix" else "+$value$suffix"
            value > 0 -> value.toString()
            suffix.isNotEmpty() -> "$value$suffix"
            else -> value.toString()
        }
    }
}

// Binary Search Tree interface and implementing classes
sealed interface BinarySearchTree {
    fun insert(value: String): BinarySearchTree
    fun contains(value: String): Boolean
    fun size(): Int
    override fun toString(): String

    object Empty : BinarySearchTree {
        override fun insert(value: String): BinarySearchTree {
            return Node(value, Empty, Empty)
        }

        override fun contains(value: String): Boolean = false

        override fun size(): Int = 0

        override fun toString(): String = "()"
    }

    data class Node(val value: String, val left: BinarySearchTree, val right: BinarySearchTree) : BinarySearchTree {
        override fun insert(newValue: String): BinarySearchTree {
            return when {
                newValue < value -> Node(value, left.insert(newValue), right)
                newValue > value -> Node(value, left, right.insert(newValue))
                else -> this
            }
        }

        override fun contains(target: String): Boolean {
            return when {
                target == value -> true
                target < value -> left.contains(target)
                else -> right.contains(target)
            }
        }

        override fun size(): Int = 1 + left.size() + right.size()

        override fun toString(): String {
            val leftStr = if (left == Empty) "" else left.toString()
            val rightStr = if (right == Empty) "" else right.toString()
            return "($leftStr$value$rightStr)"
        }
    }
}

